app/repositories/channels.py

```diff
@@ -0,0 +1,205 @@
+        "topic": topic,
+        "read_priv": read_priv,
+        "write_priv": write_priv,
+        "auto_join": auto_join,
+    }
+    rec_id = await app.state.services.database.execute(query, params)
+
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM channels
+         WHERE id = :id
+    """
+    params = {
+        "id": rec_id,
+    }
+
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return dict(rec)
+
+
+async def fetch_one(
+    id: Optional[int] = None,
+    name: Optional[str] = None,
+) -> Optional[dict[str, Any]]:
+    """Fetch a single channel."""
+    if id is None and name is None:
+        raise ValueError("Must provide at least one parameter.")
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM channels
+         WHERE id = COALESCE(:id, id)
+            AND name = COALESCE(:name, name)
+    """
+    params = {
+        "id": id,
+        "name": name,
+    }
+
+    rec = await app.state.services.database.fetch_one(query, params)
+    return dict(rec) if rec is not None else None
+
+
+async def fetch_count(
+    read_priv: Optional[int] = None,
+    write_priv: Optional[int] = None,
+    auto_join: Optional[bool] = None,
+) -> int:
+    if read_priv is None and write_priv is None and auto_join is None:
+        raise ValueError("Must provide at least one parameter.")
+
+    query = """\
+        SELECT COUNT(*) AS count
+          FROM channels
+         WHERE read_priv = COALESCE(:read_priv, read_priv)
+            AND write_priv = COALESCE(:write_priv, write_priv)
+            AND auto_join = COALESCE(:auto_join, auto_join)
+    """
+    params = {
+        "read_priv": read_priv,
+        "write_priv": write_priv,
+        "auto_join": auto_join,
+    }
+
+    rec = await app.state.services.database.fetch_one(query, params)
+    assert rec is not None
+    return rec["count"]
+
+
+async def fetch_many(
+    read_priv: Optional[int] = None,
+    write_priv: Optional[int] = None,
+    auto_join: Optional[bool] = None,
+    page: Optional[int] = None,
+    page_size: Optional[int] = None,
+) -> list[dict[str, Any]]:
+    """Fetch multiple channels from the database."""
+    query = f"""\
+        SELECT {READ_PARAMS}
+          FROM channels
+         WHERE read_priv = COALESCE(:read_priv, read_priv)
+            AND write_priv = COALESCE(:write_priv, write_priv)
+            AND auto_join = COALESCE(:auto_join, auto_join)
+    """
+    params = {
+        "read_priv": read_priv,
+        "write_priv": write_priv,
+        "auto_join": auto_join,
+    }
+
+    if page is not None and page_size is not None:
+        query += """\
+            LIMIT :limit
+           OFFSET :offset
+        """
+        params["limit"] = page_size
+        params["offset"] = (page - 1) * page_size
+
+    recs = await app.state.services.database.fetch_all(query, params)
+    return [dict(rec) for rec in recs]
+
+
+async def update(
+    name: str,
```

Comment:
this should be by id

Reply:
name support would be nice as well (channels are generally referenced by name) but for correctness let's just keep by id for now